[FLINK-10291] Generate JobGraph with fixed/configurable JobID in StandaloneJobClusterEntrypoint

yanghua opened the pull request:
## What is the purpose of the change

*This pull request generate JobGraph with fixed JobID in StandaloneJobClusterEntrypoint*

## Brief change log

  - *Generate JobGraph with fixed JobID in StandaloneJobClusterEntrypoint*

## Verifying this change

This change is already covered by existing tests, such as *StandaloneJobClusterEntryPointTest#testJobGraphRetrieval , testSavepointRestoreSettings*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

uce commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java:
```diff
@@ -67,6 +68,8 @@
 	@Nonnull
 	private final SavepointRestoreSettings savepointRestoreSettings;
 
+	private final JobID jobID = new JobID(Long.MIN_VALUE, Long.MIN_VALUE);
```
The resulting `JobID` for this is: `80000000000000008000000000000000`.

What do you think about using `new JobID(0, 0)` to give an all zero `JobID` instead? Since this is a fixed ID anyways, I think going with the simplest ID is preferable and an all zero ID might be easier to use when manually working with the REST API of a job instead of remembering to insert an `8` (which also looks very similar to a `0`).

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramUtils.java:
```diff
@@ -99,5 +102,21 @@ public static JobGraph createJobGraph(
 	}
 
+	/**
+	 * Creates a {@link JobGraph} from the given {@link PackagedProgram}.
```
Maybe add "with a random `JobID`".

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramUtils.java:
```diff
@@ -44,13 +45,15 @@
 	 * @param packagedProgram to extract the JobGraph from
```
Maybe add to the JavaDoc, that it generates a `JobGraph` with the specified `JobID`.

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java:
```diff
@@ -67,6 +68,8 @@
 	@Nonnull
 	private final SavepointRestoreSettings savepointRestoreSettings;
 
+	private final JobID jobID = new JobID(0, 0);
```
Let's make this a static variable `FIXED_JOB_ID`

tillrohrmann commented on flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java:
```diff
@@ -53,6 +55,7 @@ public void testJobGraphRetrieval() throws FlinkException {
 
 		assertThat(jobGraph.getName(), is(equalTo(TestJob.class.getCanonicalName() + "-suffix")));
 		assertThat(jobGraph.getMaximumParallelism(), is(parallelism));
+		assertEquals(jobGraph.getJobID(), new JobID(0, 0));
```
use `FIXED_JOB_ID` constant from `StandaloneJobClusterEntrypoint` for the expected value.

tillrohrmann commented on flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java:
```diff
@@ -68,5 +71,6 @@ public void testSavepointRestoreSettings() throws FlinkException {
 		final JobGraph jobGraph = jobClusterEntryPoint.retrieveJobGraph(configuration);
 
 		assertThat(jobGraph.getSavepointRestoreSettings(), is(equalTo(savepointRestoreSettings)));
+		assertEquals(jobGraph.getJobID(), new JobID(0, 0));
```
Same here.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java:
```diff
@@ -656,15 +657,24 @@ private void removeVertex(StreamNode toRemove) {
 	@SuppressWarnings("deprecation")
 	public JobGraph getJobGraph() {
+		return getJobGraph(null);
+	}
```
We could move this implementation into `StreamingPlan`.

tillrohrmann commented on flink-optimizer/src/main/java/org/apache/flink/optimizer/plan/StreamingPlan.java:
```diff
@@ -31,6 +32,8 @@
 
 	public abstract JobGraph getJobGraph();
 
+	public abstract JobGraph getJobGraph(JobID jobID);
```
Add `@Nullable` annotation to argument.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java:
```diff
@@ -656,15 +657,24 @@ private void removeVertex(StreamNode toRemove) {
+	 */
+	@SuppressWarnings("deprecation")
+	@Override
+	public JobGraph getJobGraph(JobID jobID) {
```
`@Nullable` annotation for `jobID` missing.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -86,6 +87,10 @@ public static JobGraph createJobGraph(StreamGraph streamGraph) {
 		return new StreamingJobGraphGenerator(streamGraph).createJobGraph();
 	}
 
+	public static JobGraph createJobGraph(StreamGraph streamGraph, JobID jobID) {
```
`@Nullable` annotation.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -86,6 +87,10 @@ public static JobGraph createJobGraph(StreamGraph streamGraph) {
 		return new StreamingJobGraphGenerator(streamGraph).createJobGraph();
```
This method could call `createJobGraph(streamGraph, null)`.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -108,6 +113,10 @@ public static JobGraph createJobGraph(StreamGraph streamGraph) {
+		this(streamGraph, null);
+	}
+
+	private StreamingJobGraphGenerator(StreamGraph streamGraph, JobID jobID) {
```
`@Nullable` annotation missing.

tillrohrmann commented:
Please also rebase onto the latest master if possible.

yanghua commented:
@tillrohrmann I have refactored and rebased this PR. I found the original `retrieveJobGraph ` method in `StandaloneJobClusterEntrypoint ` has been moved into `ClassPathJobGraphRetriever` (introduced 
 in PR #6743 ). So I moved my changes into the `ClassPathJobGraphRetriever#retrieveJobGraph`. Hope it matches your expect.
